File: airpoll-backend/src/main/java/it/airpoll/service/AirpollService.java
package it.airpoll.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Profile;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import it.airpoll.common.AirpollCommonService;
import it.airpoll.common.exception.ItemNotFoundException;
import it.airpoll.dto.AirpollDto;
import it.airpoll.mapper.AirpollMapper;
import it.airpoll.model.City;
import it.airpoll.model.Country;
import it.airpoll.repository.CityRepository;
import it.airpoll.repository.CountryRepository;
import it.airpoll.repository.specification.CitySpecification;
import it.airpoll.repository.specification.CountrySpecification;

/**
 * @author Giacomo
 * 
 * Airpoll Service loaded in default mode. It has two overloaded method who retrieve whole data, by country, by city, by both two params. The
 * Country should be mapped by properly mapper to iterate and flat the data.
 * 
 */

@Profile("default")
@Service
public class AirpollService extends AirpollCommonService<Country, AirpollDto, Integer, CountryRepository, AirpollMapper> {
	
	@Autowired protected CityRepository cityRepository;

	public AirpollService(CountryRepository repository, AirpollMapper mapper) {
		super(repository, mapper);
	}
	
	@Transactional(readOnly = true)
	public List<AirpollDto> get(Integer countryId, Integer page) throws ItemNotFoundException {
		Specification<Country> spec = Specification.where(CountrySpecification.findByCountry(countryId));
		
		Page<Country> entities = repository.findAll(spec, PageRequest.of(0, page));
		if (entities.isEmpty()) throw new ItemNotFoundException("Entities not found!");
		return mapper.toDtoS(entities.toList());
	}
	
	@Transactional(readOnly = true)
	public List<AirpollDto> get(Integer countryId, Integer cityId, Integer page) throws ItemNotFoundException {
		Specification<City> spec = Specification.where(CitySpecification.findByCountry(countryId)
				.and(CitySpecification.findByCity(cityId)));
		
		Page<City> entities = cityRepository.findAll(spec, PageRequest.of(0, page));
		if (entities.isEmpty()) throw new ItemNotFoundException("Entities not found!");
		return mapper.toDTOS(entities.toList());
	}
}

File: airpoll-frontend/src/app/services/airpoll.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http'
import { Observable } from 'rxjs';
import { AirpollData } from '../model/airpollData';
import { Country } from '../model/country';
import { City } from '../model/city';

@Injectable({
  providedIn: 'root'
})
export class AirpollService {

  constructor(private http: HttpClient) { }

  list(countryId: Number, cityId: Number, page: Number): Observable<AirpollData[]> {
    let params = new HttpParams();
    params = params.append('countryId', !countryId ? '' : countryId.toString());
    params = params.append('cityId', !cityId ? '' : cityId.toString());
    params = params.append('page', page.toString());
    return this.http.get<AirpollData[]>('data', { params: params });
  }

  countryList(): Observable<Country[]> {
    return this.http.get<Country[]>('country');
  }

  cityList(): Observable<City[]> {
    return this.http.get<City[]>('city');
  }
}
